### ISSUE TITLE
Show zero coverage for unreachable blocks (e.g., dropped after const evaluation)

### ISSUE THREAD

**Issue by richkadel** [rust-lang/rust#84018]
Source-based code coverage (via `-Z instrument-coverage`) does not show coverage for unreachable blocks that were optimized out during MIR transformations. For example:

```rust
if false {
    println!("never executed");
}
```

Ideally the second line would show a coverage execution count of `0`, but, assuming this snippet is executed, it would instead show something like:

```
1   if false {
        println!("never executed");
1   }
```

The coverage percentage for this block would be computed as 100%, but should be about 66% (by lines covered).

The MIR `InstrumentCoverage` pass is performed before MIR optimizations, so the coverage instrumentation is available in MIR for line 2. However, subsequent optimizations drop the MIR block, including the coverage `Statement` before the codegen phase. The coverage map is generated during codegen, and can use `tcx` queries to extract data from the `optimized_mir()`, but that version of the MIR has already been purged of the unreachable blocks.

As far as I can tell, `tcx` queries for intermediate MIR representations (for example, `tcx.mir_promoted()`) are not available at codegen (they have already been "stolen" by the optimization passes).

To solve this problem, I believe I need to cache at least the coverage statements (or their data at least) for the pre-optimized MIR, before those statements are potentially dropped by subsequent MIR passes.

cc: @tmandry @wesleywiser

**Comment by richkadel** [comment]
I think the easiest way to resolve this would be to just disable simplification of unreachable blocks (due to const eval), if `-Zinstrument-coverage`.

Is there an option for that? I haven't looked yet.

**Comment by richkadel** [comment]
> I think the easiest way to resolve this would be to just disable simplification of unreachable blocks (due to const eval), if `-Zinstrument-coverage`.

Replying to my own question:

This would not have worked. I noticed (and found in-code comments to confirm) that dead code cannot be left in the MIR.

But I solved this in a better way (see PR #84797) by simply saving coverage code regions from dropped `BasicBlocks` as `Unreachable` coverage, adding them to the `START_BLOCK`. This seems to work really well.
